src/ai/backend/storage/cephfs/__init__.py

```diff
@@ -0,0 +1,103 @@
+import asyncio
+import os
+from typing import Dict, List
+from uuid import UUID
+
+from ai.backend.common.types import BinarySize
+
+from ..exception import ExecutionError
+from ..types import FSUsage, Optional, VFolderCreationOptions
+from ..vfs import BaseVolume
+
+
+async def read_file(loop: asyncio.AbstractEventLoop, filename: str) -> str:
+    def _read():
+        with open(filename, "r") as fr:
+            return fr.read()
+
+    return await loop.run_in_executor(None, _read())
+
+
+async def write_file(loop: asyncio.AbstractEventLoop, filename: str, contents: str, perm="w"):
+    def _write():
+        with open(filename, perm) as fw:
+            fw.write(contents)
+
+    await loop.run_in_executor(None, _write())
+
+
+async def run(cmd: str) -> str:
+    proc = await asyncio.create_subprocess_shell(
+        cmd, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE
+    )
+    out, err = await proc.communicate()
+    if err:
+        raise ExecutionError(err.decode())
+    return out.decode()
+
+
+class CephFSVolume(BaseVolume):
+    loop: asyncio.AbstractEventLoop
+    registry: Dict[str, int]
+    project_id_pool: List[int]
+
+    async def init(self) -> None:
+        available = True
+        try:
+            await asyncio.create_subprocess_exec(
+                b"ceph",
+                b"--version",
+                stdout=asyncio.subprocess.PIPE,
+                stderr=asyncio.subprocess.STDOUT,
+            )
+        except FileNotFoundError:
+            available = False
+
+        if not available:
+            raise RuntimeError("Ceph is not installed. ")
+
+    # ----- volume operations -----
+    async def create_vfolder(
+        self, vfid: UUID, options: Optional[VFolderCreationOptions], exist_ok: bool = False
+    ) -> None:
+        vfpath = self.mangle_vfpath(vfid)
+        loop = asyncio.get_running_loop()
+        await loop.run_in_executor(
+            None,
+            lambda: vfpath.mkdir(0o755, parents=True, exist_ok=exist_ok),
+        )
+        if not (options is None or options.quota is None or options.quota == 0):
+            quota = options.quota
+            await self.set_quota(vfpath, quota)
+
+    async def get_fs_usage(self) -> FSUsage:
+        stat = await run(f"df -h {self.mount_path} | grep {self.mount_path}")
```

Comment:
Parsing output of command result should always be considered as last resort since it depends on lots of other things ($PATH setup, output format, ...). I think we can replace `df -h` with `os.statvfs()` API, which is provided by Python's built-in `os` library.

Reply:
There is a need in better  Python command to display more clear file system usage.
df -h /mnt/vfroot/ceph-fuse command gives following output:
```
Filesystem      Size  Used Avail Use% Mounted on
ceph-fuse       5.7G     0  5.7G   0% /mnt/vfroot/ceph-fuse
```
The os.statvfs gives following output:
```
os.statvfs("/mnt/vfroot/ceph-fuse")
os.statvfs_result(f_bsize=4194304, f_frsize=4194304, f_blocks=1450, f_bfree=1450, f_bavail=1450, f_files=79, f_ffree=0, f_favail=0, f_flag=4102, f_namemax=255)
```